shell/platform/android/surface/android_surface_transaction.cc

```diff
@@ -0,0 +1,76 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/android/surface/android_surface_transaction.h"
+
+#include "flutter/fml/native_library.h"
+
+namespace flutter {
+
+AndroidSurfaceTransaction::AndroidSurfaceTransaction()
+    : pending_transaction_(nullptr) {
+#if FML_OS_ANDROID
+  auto libandroid = fml::NativeLibrary::Create("libandroid.so");
+  FML_DCHECK(libandroid);
+  ASurfaceTransaction_create =
+      libandroid
+          ->ResolveFunction<ASurfaceTransaction_create_FPN>(
+              "ASurfaceTransaction_create")
+          .value_or(nullptr);
+  ASurfaceTransaction_setFrameTimeline =
+      libandroid
+          ->ResolveFunction<ASurfaceTransaction_setFrameTimeline_FPN>(
+              "ASurfaceTransaction_setFrameTimeline")
+          .value_or(nullptr);
+  ASurfaceTransaction_apply =
+      libandroid
+          ->ResolveFunction<ASurfaceTransaction_apply_FPN>(
+              "ASurfaceTransaction_apply")
+          .value_or(nullptr);
+  ASurfaceTransaction_delete =
+      libandroid
+          ->ResolveFunction<ASurfaceTransaction_delete_FPN>(
+              "ASurfaceTransaction_delete")
+          .value_or(nullptr);
+  if (!ASurfaceTransaction_create || !ASurfaceTransaction_setFrameTimeline ||
+      !ASurfaceTransaction_apply || !ASurfaceTransaction_delete) {
+    ASurfaceTransaction_create = nullptr;
+    ASurfaceTransaction_setFrameTimeline = nullptr;
+    ASurfaceTransaction_apply = nullptr;
+    ASurfaceTransaction_delete = nullptr;
+  }
+#else
+  ASurfaceTransaction_create = nullptr;
+  ASurfaceTransaction_setFrameTimeline = nullptr;
+  ASurfaceTransaction_apply = nullptr;
+  ASurfaceTransaction_delete = nullptr;
+#endif  // FML_OS_ANDROID
+}
+
+void AndroidSurfaceTransaction::Begin() {
+  FML_DCHECK(!pending_transaction_);
+  if (ASurfaceTransaction_create) {
+    pending_transaction_ = ASurfaceTransaction_create();
+  }
+}
+
+void AndroidSurfaceTransaction::SetVsyncId(int64_t vsync_id) {
+  if (ASurfaceTransaction_create) {
+    FML_DCHECK(pending_transaction_);
+    if (vsync_id != 0) {
```

Comment:
we should have a named constant for this invalid vsync id

Reply:
Sure, where do you think I should declare it? Not confident about any namespacing/importing rules in this repo, and this would be used in:

- `flow/surface_frame.h`
- `shell/gpu/gpu_surface_gl_delegate.h`
- `shell/platform/android/external_view_embedder/external_view_embedder.cc`
- `shell/platform/android/surface/android_surface_transaction.cc`

Reply:
Added to `common/constants.h`